src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,201 @@
+                outEvent.EventType == TransferSession::OutputEventType::kStatusReceived)
+            {
+                CleanUp();
+                return err;
+            }
+        }
+        mTransfer.GetNextAction(outEvent);
+    };
+    return err;
+}
+
+void AsyncTransferFacilitator::OnExchangeClosing(Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "OnExchangeClosing, ec: " ChipLogFormatExchange, ChipLogValueExchange(ec));
+    mExchange.Release();
+}
+
+void AsyncTransferFacilitator::OnResponseTimeout(Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "OnResponseTimeout, ec: " ChipLogFormatExchange, ChipLogValueExchange(ec));
+    CleanUp();
+}
+
+CHIP_ERROR AsyncResponder::PrepareForTransfer(Messaging::ExchangeContext * exchangeCtx, TransferRole role,
+                                              BitFlags<TransferControlFlags> xferControlOpts, uint16_t maxBlockSize,
+                                              System::Clock::Timeout timeout)
+{
+    VerifyOrReturnError(exchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(!mExchange, CHIP_ERROR_INCORRECT_STATE);
+
+    mTimeout = timeout;
+
+    ReturnErrorOnFailure(mTransfer.WaitForTransfer(role, xferControlOpts, maxBlockSize, mTimeout));
+
+    mExchange.Grab(exchangeCtx);
+
+    return CHIP_NO_ERROR;
+}
+
+bdx::StatusCode AsyncResponder::GetBdxStatusCodeFromChipError(CHIP_ERROR err)
+{
+    if (err == CHIP_ERROR_INCORRECT_STATE)
+    {
+        return bdx::StatusCode::kUnexpectedMessage;
+    }
+    if (err == CHIP_ERROR_INVALID_ARGUMENT)
+    {
+        return bdx::StatusCode::kBadMessageContents;
+    }
+    return bdx::StatusCode::kUnknown;
+}
+
+void AsyncResponder::NotifyEventHandled(CHIP_ERROR eventHandlingResult)
+{
+    // We can only come here after we have handled a BDX message (like ReceiveInit/BlockQuery) asynchronously
+    // in the subclass. The base class needs to notify us when the handling of the event is done and the event
+    // handling result - success or failure.
+    ChipLogDetail(BDX, "NotifyEventHandled : error %" CHIP_ERROR_FORMAT, eventHandlingResult.Format());
+
+    // If there was an error, we abort the transfer.
+    if (eventHandlingResult != CHIP_NO_ERROR)
+    {
+        CleanUp();
+    }
+    else
+    {
+
+        // We need to get the next output event from the state machine which is a response to the
+        // BDX message handled by the subclass (ReceiveAccept/Block).
+        TransferSession::OutputEvent outEvent;
+
+        mTransfer.GetNextAction(outEvent);
+        while (outEvent.EventType != TransferSession::OutputEventType::kNone)
+        {
+            if (outEvent.EventType == TransferSession::OutputEventType::kMsgToSend)
+            {
+                AsyncTransferFacilitator::SendMessage(outEvent);
+                mTransfer.GetNextAction(outEvent);
+            }
+            else
+            {
+                break;
```

Comment:
So if we got some other event here, that's not "send a message"... we will just drop it on the floor and not do anything at all with it?  Who will process that event?  When?  Why?

Again, it feels like we need to have a _very_ clear idea here of who is doing what control flow bits at what points.  It feels like we have a setup that kind of works on the happy paths and as soon as there is any error no one knows who's responsible for handling it and how.

Reply:
This is used to handle the asynchronous part as described in the comment above. This needs more work in relation to https://github.com/project-chip/connectedhomeip/pull/29364/files#r1343191242